sql/sql_table.cc

```diff
@@ -6820,6 +6820,14 @@ bool mysql_alter_table(THD *thd,char *new_db, char *new_name,
   else
     create_info->data_file_name=create_info->index_file_name=0;
 
+  if (old_db_type != new_db_type && old_db_type->db_type == DB_TYPE_HEAP)
```

Comment:
Looks like leaky abstraction here - is it possible to make the change in the MEMORY SE layer instead?

Reply:
I tried to do that in a few different ways, but couldn't do it. It would require a before/during alter hook in the (old) storage engine (so the memory se would be able to change the key_block_size when the engine is changed), and there isn't such a thing in the code - only the validation step, with the new storage engine, where innodb raises the warnings/errors.

Reply:
Ack, then only the testcase comment needs addressing

Reply:
Correction: looks like the change which is required for PS-3943 also fixes this issue, without this change here. I updated my commit based on that, and I'll also update the test cases tomorrow.